### PR TITLE
THRIFT-3357: enum_containers option to generate EnumSet/EnumMap

### PR DESCRIPTION
**Pull request by afds**
Additional java compiler option to generate EnumSet/EnumMap instead
of HashSet/HashMap where elements/keys are enums.

### CODE REVIEW

**Comment by afds** [comment]
https://issues.apache.org/jira/browse/THRIFT-3357

File: compiler/cpp/src/generate/t_java_generator.cc [hunk 1]
```diff
@@ -5139,8 +5215,11 @@ THRIFT_REGISTER_GENERATOR(
     "    reuse-objects:   Data objects will not be allocated, but existing instances will be used "
     "(read and write).\n"
     "    sorted_containers:\n"
-    "                     Use TreeSet/TreeMap instead of HashSet/HashMap as a implementation of "
+    "                     Use TreeSet/TreeMap instead of HashSet/HashMap as an implementation of "
     "set/map.\n"
+    "    enum_containers:\n"
```
Comment (Jens-G) [inline compiler/cpp/src/generate/t_java_generator.cc]: Not so sure if we need a new option at all. The default assumptions for Thrift containers are described at https://thrift.apache.org/docs/types . If the patch does not break compatibility and on top is more efficient, why not make it the default? $0,02

**Comment by afds** [comment]
I'm perfectly fine to make it default.

Theoretically possible issues is that someone heavily relies on HashMap/HashSet or uses _null_ keys/elements (EnumMap/EnumSet does not allow _null_ keys/elements). But in that case, we indeed, need to make it default to punish them.

**Comment by afds** [comment]
Created a new pull request here: https://github.com/apache/thrift/pull/1253

**Pull request closed without merging**